File: _code-samples/confidential-transfers/py/send_confidential_payments.py
import asyncio
import json
import os

from xrpl.clients import JsonRpcClient
from xrpl.ext.confidential import (
    MPTCrypto,
    prepare_confidential_merge_inbox,
    prepare_confidential_send,
)
from xrpl.models import Batch, LedgerEntry, Tx
from xrpl.models.requests.ledger_entry import MPToken
from xrpl.models.transactions import ConfidentialMPTSend, Transaction
from xrpl.models.transactions.batch import BatchFlag
from xrpl.models.transactions.transaction import TransactionFlag
from xrpl.transaction import (
    autofill,
    combine_batch_signers,
    sign_multiaccount_batch,
    submit_and_wait,
)
from xrpl.wallet import Wallet, generate_faucet_wallet

from confidential_transfers_setup import main as run_setup

# Connect to the network ----------------------
client = JsonRpcClient("https://s.devnet.rippletest.net:51234")
crypto = MPTCrypto()

EXPLORER = "https://devnet.xrpl.org"

# Load setup data ----------------------
# This step checks for the necessary setup data to run the tutorial.
# If missing, confidential_transfers_setup.py will generate it.
if not os.path.exists("confidential_transfers_setup.json"):
    print("\n=== Setup data doesn't exist. Running setup script... ===\n")
    asyncio.run(run_setup())

with open("confidential_transfers_setup.json") as setup_file:
    setup_data = json.load(setup_file)

fund = setup_data["fund"]
stablecoin = setup_data["stablecoin"]

# Set up accounts
print("\n=== Getting accounts... ===")
orchestrator = generate_faucet_wallet(client)
seller = Wallet.from_seed(setup_data["seller"]["seed"])
buyer = Wallet.from_seed(setup_data["buyer"]["seed"])
auditor = Wallet.from_seed(setup_data["auditor"]["seed"])

# A confidential encryption keypair is generated at random and can't be rebuilt
# from an account seed, so the setup script saved each one it created.
seller_privkey = setup_data["seller"]["privateKey"]
seller_pubkey = setup_data["seller"]["publicKey"]
buyer_privkey = setup_data["buyer"]["privateKey"]
buyer_pubkey = setup_data["buyer"]["publicKey"]
issuer_pubkey = setup_data["issuer"]["publicKey"]
auditor_privkey = setup_data["auditor"]["privateKey"]
auditor_pubkey = setup_data["auditor"]["publicKey"]

print(f"Orchestrator address: {orchestrator.address}")
print(f"Seller address: {seller.address}")
print(f"Buyer address: {buyer.address}")
print(f"Auditor address: {auditor.address}")

# Make each holder's balance spendable ----------------------
print("\n=== Merging inbox balance for seller and buyer... ===")
tokens = (fund, stablecoin)
holders = (
    ("Seller", seller, seller_privkey),
    ("Buyer", buyer, buyer_privkey),
)
holdings = ((fund, seller), (stablecoin, buyer))

for token, holder in holdings:
    merge_tx = prepare_confidential_merge_inbox(
        client, holder, token["mptIssuanceID"]
    )
    merge_response = submit_and_wait(merge_tx, client, holder, autofill=True)
    merge_result = merge_response.result["meta"]["TransactionResult"]
    if merge_result != "tesSUCCESS":
        print(f"Error: Unable to merge the {token['ticker']} inbox: {merge_result}")
        exit(1)
    print(f"{holder.address} holds spendable confidential {token['ticker']}.")
    print(f"{EXPLORER}/transactions/{merge_response.result['hash']}\n")

# Build both confidential payments ----------------------
fund_amount = 100
cash_amount = 500

fund_payment = prepare_confidential_send(
    client,
    seller,
    buyer.address,
    fund["mptIssuanceID"],
    fund_amount,
    seller_privkey,
    seller_pubkey,
    buyer_pubkey,
    issuer_pubkey,
    auditor_pubkey,
)
cash_payment = prepare_confidential_send(
    client,
    buyer,
    seller.address,
    stablecoin["mptIssuanceID"],
    cash_amount,
    buyer_privkey,
    buyer_pubkey,
    seller_pubkey,
    issuer_pubkey,
    auditor_pubkey,
)

print("=== Prepared confidential payments ===")
print(f"Payment1 (Fund): {json.dumps(fund_payment.to_xrpl(), indent=2)}")
print(f"\nPayment2 (Cash): {json.dumps(cash_payment.to_xrpl(), indent=2)}")

# Read the fees the payments carried before overwriting them, because the outer
# Batch needs them later on.
inner_fees = int(fund_payment.fee) + int(cash_payment.fee)

# Every inner Batch transaction needs the tfInnerBatchTxn flag and a Fee of 0, so
# set the flag and overwrite the Fee the helper function filled in.
fund_payment, cash_payment = [
    ConfidentialMPTSend.from_dict(
        {**payment.to_dict(), "fee": "0", "flags": TransactionFlag.TF_INNER_BATCH_TXN}
    )
    for payment in (fund_payment, cash_payment)
]

# Settle both payments atomically ----------------------
print("\n=== Submit confidential payments in batch... ===")
print(f"Seller sends {fund['ticker']} to Buyer.")
print(f"Buyer sends {stablecoin['ticker']} to Seller.\n")

batch_tx = Batch(
    account=orchestrator.address,
    flags=BatchFlag.TF_ALL_OR_NOTHING,
    raw_transactions=[fund_payment, cash_payment],
)

autofilled_batch_tx = autofill(batch_tx, client, 2)
autofilled_batch_tx = Batch.from_dict(
    {
        **autofilled_batch_tx.to_dict(),
        "fee": str(int(autofilled_batch_tx.fee) + inner_fees),
    }
)

seller_batch = sign_multiaccount_batch(seller, autofilled_batch_tx)
buyer_batch = sign_multiaccount_batch(buyer, autofilled_batch_tx)
combined_batch_tx = combine_batch_signers([seller_batch, buyer_batch])

batch_response = submit_and_wait(combined_batch_tx, client, orchestrator)
if batch_response.result["meta"]["TransactionResult"] != "tesSUCCESS":
    result_code = batch_response.result["meta"]["TransactionResult"]
    print(f"Error: Unable to submit the Batch: {result_code}")
    exit(1)
print(f"Batch transaction hash: {batch_response.result['hash']}")
print(f"{EXPLORER}/transactions/{batch_response.result['hash']}")

# Verify each payment individually ----------------------
# A tesSUCCESS on the Batch only means the Batch itself was well-formed.
# Hash each inner transaction and look it up to confirm both payments applied.
print("\n=== Verifying both payments... ===")
raw_transactions = batch_response.result["tx_json"]["RawTransactions"]
for index, raw_transaction in enumerate(raw_transactions):
    inner = Transaction.from_xrpl(raw_transaction["RawTransaction"])
    inner_hash = inner.get_hash()
    inner_result = client.request(Tx(transaction=inner_hash)).result["meta"][
        "TransactionResult"
    ]
    print(f"Payment {index + 1}: {inner_result}")
    print(f"{EXPLORER}/transactions/{inner_hash}\n")

    if inner_result != "tesSUCCESS":
        print(f"Error: An inner payment failed: {inner_result}")
        exit(1)
print("Payments both successful!")

# Merge the received amounts into each spending balance ----------------------
print("\n=== Merging settled amounts into spending balance... ===")
settlements = ((fund, buyer), (stablecoin, seller))

for token, recipient in settlements:
    merge_tx = prepare_confidential_merge_inbox(
        client, recipient, token["mptIssuanceID"]
    )
    merge_response = submit_and_wait(merge_tx, client, recipient, autofill=True)
    merge_result = merge_response.result["meta"]["TransactionResult"]
    if merge_result != "tesSUCCESS":
        print(f"Error: Unable to merge the {token['ticker']} inbox: {merge_result}")
        exit(1)
    print(f"{recipient.address} can spend the {token['ticker']} it received.")
    print(f"{EXPLORER}/transactions/{merge_response.result['hash']}\n")

# Decrypt balances as each holder ----------------------
confidential_supplies = {}
for token in tokens:
    issuance = client.request(
        LedgerEntry(mpt_issuance=token["mptIssuanceID"]),
    ).result["node"]
    confidential_supplies[token["mptIssuanceID"]] = int(
        issuance["ConfidentialOutstandingAmount"]
    )

print("=== Decrypting balances as each holder... ===")
for name, holder, holder_privkey in holders:
    print(f"\n{name} reads its own balance as:")
    for token in tokens:
        mptoken = client.request(
            LedgerEntry(
                mptoken=MPToken(
                    mpt_issuance_id=token["mptIssuanceID"], account=holder.address
                ),
            ),
        ).result["node"]
        spending_balance = mptoken["ConfidentialBalanceSpending"]
        balance = crypto.decrypt(
            holder_privkey,
            spending_balance[:66],
            spending_balance[66:],
            range_high=confidential_supplies[token["mptIssuanceID"]],
        )
        print(f"     - {balance} {token['ticker']}\n")

# Decrypt the balances and amounts as the auditor ----------------------
print("\n=== Decrypting balances and amounts as the auditor... ===")
for name, holder, _ in holders:
    print(f"Auditor reads the {name.lower()}'s balance as:")
    for token in tokens:
        mptoken = client.request(
            LedgerEntry(
                mptoken=MPToken(
                    mpt_issuance_id=token["mptIssuanceID"], account=holder.address
                ),
            ),
        ).result["node"]
        auditor_balance = mptoken["AuditorEncryptedBalance"]
        auditor_view = crypto.decrypt(
            auditor_privkey,
            auditor_balance[:66],
            auditor_balance[66:],
            range_high=confidential_supplies[token["mptIssuanceID"]],
        )
        print(f"     - {auditor_view} {token['ticker']}")

print("\nAuditor reads the settled amounts as:")
settled = ((fund, fund_payment), (stablecoin, cash_payment))

for token, payment in settled:
    auditor_amount = payment.auditor_encrypted_amount
    settled_amount = crypto.decrypt(
        auditor_privkey,
        auditor_amount[:66],
        auditor_amount[66:],
        range_high=confidential_supplies[token["mptIssuanceID"]],
    )
    print(f"     - {settled_amount} {token['ticker']}")

File: _code-samples/confidential-transfers/js/sendConfidentialPayments.js
import fs from 'fs'
import {
  BatchFlags,
  Client,
  GlobalFlags,
  Wallet,
  combineBatchSigners,
  deriveConfidentialKeypair,
  fetchMPToken,
  fetchMPTokenIssuance,
  getConfidentialBalance,
  hashes,
  loadMptCrypto,
  prepareConfidentialMergeInbox,
  prepareConfidentialSend,
  signMultiBatch,
  validate
} from 'xrpl'

import { setup } from './confidentialTransfersSetup.js'

// Connect to the network ----------------------
const client = new Client('wss://s.devnet.rippletest.net:51233')
await client.connect()

const EXPLORER = 'https://devnet.xrpl.org'

// Load setup data ----------------------
// This step checks for the necessary setup data to run the tutorial.
// If missing, confidentialTransfersSetup.js will generate it.
if (!fs.existsSync('confidentialTransfersSetup.json')) {
  console.log(`\n=== Setup data doesn't exist. Running setup script... ===\n`)
  await setup()
}

const setupData = JSON.parse(
  fs.readFileSync('confidentialTransfersSetup.json', 'utf8')
)

// Set up accounts
console.log(`\n=== Getting accounts... ===`)
const { wallet: orchestrator } = await client.fundWallet()
const seller = Wallet.fromSeed(setupData.seller.seed)
const buyer = Wallet.fromSeed(setupData.buyer.seed)
const auditor = Wallet.fromSeed(setupData.auditor.seed)

// deriveConfidentialKeypair rebuilds a confidential encryption keypair from an
// account seed. The same seed always gives the same keypair.
const sellerKeys = deriveConfidentialKeypair(seller.seed)
const buyerKeys = deriveConfidentialKeypair(buyer.seed)
const auditorKeys = deriveConfidentialKeypair(auditor.seed)

console.log(`Orchestrator address: ${orchestrator.address}`)
console.log(`Seller address: ${seller.address}`)
console.log(`Buyer address: ${buyer.address}`)
console.log(`Auditor address: ${auditor.address}`)

// Make each holder's balance spendable ----------------------
console.log(`\n=== Merging inbox balance for seller and buyer... ===`)
const tokens = [setupData.fund, setupData.stablecoin]
const holders = [
  { name: 'Seller', holder: seller, keys: sellerKeys },
  { name: 'Buyer', holder: buyer, keys: buyerKeys }
]
const holdings = [
  { ...setupData.fund, holder: seller },
  { ...setupData.stablecoin, holder: buyer }
]

for (const { ticker, mptIssuanceID, holder } of holdings) {
  const mergeTx = await prepareConfidentialMergeInbox(client, {
    account: holder.address,
    mptIssuanceID
  })
  const mergeResponse = await client.submitAndWait(mergeTx, {
    wallet: holder,
    autofill: true
  })

  const mergeResult = mergeResponse.result.meta.TransactionResult
  if (mergeResult !== 'tesSUCCESS') {
    console.error(`Error: Unable to merge the ${ticker} inbox:`, mergeResult)
    await client.disconnect()
    process.exit(1)
  }
  console.log(`${holder.address} holds spendable confidential ${ticker}.`)
  console.log(`${EXPLORER}/transactions/${mergeResponse.result.hash}\n`)
}

// Build both confidential payments ----------------------
const fundAmount = BigInt(100)
const cashAmount = BigInt(500)

const [fundPayment, cashPayment] = await Promise.all([
  prepareConfidentialSend(client, {
    account: seller.address,
    destination: buyer.address,
    mptIssuanceID: setupData.fund.mptIssuanceID,
    amount: fundAmount,
    senderKeypair: sellerKeys
  }),
  prepareConfidentialSend(client, {
    account: buyer.address,
    destination: seller.address,
    mptIssuanceID: setupData.stablecoin.mptIssuanceID,
    amount: cashAmount,
    senderKeypair: buyerKeys
  })
])

console.log(`=== Prepared confidential payments ===`)
console.log(`Payment1 (Fund): ${JSON.stringify(fundPayment, null, 2)}`)
console.log(`\nPayment2 (Cash): ${JSON.stringify(cashPayment, null, 2)}`)

// Every inner batch transaction must have the tfInnerBatchTxn flag set.
fundPayment.Flags = GlobalFlags.tfInnerBatchTxn
cashPayment.Flags = GlobalFlags.tfInnerBatchTxn

// Settle both payments atomically ----------------------
console.log(`=== Submit confidential payments in batch... ===`)
console.log(`Seller sends ${setupData.fund.ticker} to Buyer.`)
console.log(`Buyer sends ${setupData.stablecoin.ticker} to Seller.\n`)

const batchTx = {
  TransactionType: 'Batch',
  Account: orchestrator.address,
  Flags: BatchFlags.tfAllOrNothing,
  RawTransactions: [
    { RawTransaction: fundPayment },
    { RawTransaction: cashPayment }
  ]
}
validate(batchTx)

const autofilledBatchTx = await client.autofill(batchTx, 2)

const sellerBatch = { ...autofilledBatchTx }
signMultiBatch(seller, sellerBatch)
const buyerBatch = { ...autofilledBatchTx }
signMultiBatch(buyer, buyerBatch)
const combinedBatchTx = combineBatchSigners([sellerBatch, buyerBatch])

const batchResponse = await client.submitAndWait(combinedBatchTx, {
  wallet: orchestrator
})
if (batchResponse.result.meta.TransactionResult !== 'tesSUCCESS') {
  const resultCode = batchResponse.result.meta.TransactionResult
  console.error('Error: Unable to submit the Batch:', resultCode)
  await client.disconnect()
  process.exit(1)
}
console.log(`Batch transaction hash: ${batchResponse.result.hash}`)
console.log(`${EXPLORER}/transactions/${batchResponse.result.hash}`)

// Verify each payment individually ----------------------
// A tesSUCCESS on the Batch only means the Batch itself was well-formed. 
// Hash each inner transaction and look it up to confirm both payments applied.
console.log(`\n=== Verifying both payments... ===`)
const rawTransactions = batchResponse.result.tx_json.RawTransactions

for (const [index, { RawTransaction }] of rawTransactions.entries()) {
  const innerHash = hashes.hashSignedTx(RawTransaction)
  const innerTx = await client.request({ command: 'tx', transaction: innerHash })
  const innerResult = innerTx.result.meta.TransactionResult
  console.log(`Payment ${index + 1}: ${innerResult}`)
  console.log(`${EXPLORER}/transactions/${innerHash}`)

  if (innerResult !== 'tesSUCCESS') {
    console.error('Error: An inner payment failed:', innerResult)
    await client.disconnect()
    process.exit(1)
  }
}
console.log(`\nPayments both successful!`)

// Merge the received amounts into each spending balance ----------------------
console.log(`\n=== Merging settled amounts into spending balance... ===`)
const settlements = [
  { ...setupData.fund, recipient: buyer },
  { ...setupData.stablecoin, recipient: seller }
]

for (const { ticker, mptIssuanceID, recipient } of settlements) {
  const mergeTx = await prepareConfidentialMergeInbox(client, {
    account: recipient.address,
    mptIssuanceID
  })
  const mergeResponse = await client.submitAndWait(mergeTx, {
    wallet: recipient,
    autofill: true
  })
  const mergeResult = mergeResponse.result.meta.TransactionResult
  if (mergeResult !== 'tesSUCCESS') {
    console.error(`Error: Unable to merge the ${ticker} inbox:`, mergeResult)
    await client.disconnect()
    process.exit(1)
  }
  console.log(`${recipient.address} can spend the ${ticker} it received.`)
  console.log(`${EXPLORER}/transactions/${mergeResponse.result.hash}\n`)
}

// Decrypt balances as each holder ----------------------
console.log(`=== Decrypting balances as each holder... ===`)
for (const { name, holder, keys } of holders) {
  console.log(`${name} reads its own balance as:`)
  for (const token of tokens) {
    const balance = await getConfidentialBalance(
      client,
      holder.address,
      token.mptIssuanceID,
      keys.privateKey
    )
    console.log(`     - ${balance} ${token.ticker}`)
  }
}

// Decrypt the balances and amounts as the auditor ----------------------
console.log(`\n=== Decrypting balances and amounts as the auditor... ===`)
const crypto = await loadMptCrypto()

const confidentialSupplies = {}
for (const token of tokens) {
  const issuance = await fetchMPTokenIssuance(client, token.mptIssuanceID)
  confidentialSupplies[token.mptIssuanceID] = BigInt(
    issuance.ConfidentialOutstandingAmount
  )
}

for (const { name, holder } of holders) {
  console.log(`Auditor reads the ${name.toLowerCase()}'s balance as:`)
  for (const token of tokens) {
    const mptoken = await fetchMPToken(
      client,
      holder.address,
      token.mptIssuanceID
    )
    const auditorView = await crypto.decryptAmount(
      mptoken.AuditorEncryptedBalance,
      auditorKeys.privateKey,
      confidentialSupplies[token.mptIssuanceID]
    )
    console.log(`     - ${auditorView} ${token.ticker}`)
  }
}

console.log(`\nAuditor reads the settled amounts as:`)
const settled = [
  { token: setupData.fund, payment: fundPayment },
  { token: setupData.stablecoin, payment: cashPayment }
]
for (const { token, payment } of settled) {
  const settledAmount = await crypto.decryptAmount(
    payment.AuditorEncryptedAmount,
    auditorKeys.privateKey,
    confidentialSupplies[token.mptIssuanceID]
  )
  console.log(`     - ${settledAmount} ${token.ticker}`)
}

await client.disconnect()
